tests/e2e/project.spec.ts

```diff
@@ -0,0 +1,337 @@
+
+  await page.getByText("src/").click();
+  await page.getByText("true.c").click();
+
+  await expect(page.getByText("return")).toHaveCSS(
+    "color",
+    "rgb(180, 142, 173)",
+  );
+});
+
+test("navigate deep file hierarchies", async ({ page }) => {
+  await page.goto(sourceBrowsingFixture);
+
+  const sourceTree = page.locator(".source-tree");
+
+  await sourceTree.getByText("deep/").click();
+  await sourceTree.getByText("directory/").click();
+  await sourceTree.getByText("hierarchy/").click();
+  await sourceTree.getByText("is/").click();
+  await sourceTree.getByText("entirely/").click();
+  await sourceTree.getByText("possible/").click();
+  await sourceTree.getByText("in/").nth(1).click();
+  await sourceTree.getByText("git/").click();
+  await sourceTree.getByText("repositories/").click();
+  await sourceTree.getByText(".gitkeep").click();
+  await expect(
+    page.locator("text=0801ace Add a deeply nested directory tree"),
+  ).toBeVisible();
+
+  // After a page reload the tree browser is still expanded and we're still
+  // showing the .gitkeep file.
+  {
+    await page.reload();
+
+    const sourceTree = page.locator(".source-tree");
+
+    await expect(sourceTree.getByText("deep/")).toBeVisible();
+    await expect(sourceTree.getByText("directory/")).toBeVisible();
+    await expect(sourceTree.getByText("hierarchy/")).toBeVisible();
+    await expect(sourceTree.getByText("is/")).toBeVisible();
+    await expect(sourceTree.getByText("entirely/")).toBeVisible();
+    await expect(sourceTree.getByText("possible/")).toBeVisible();
+    await expect(sourceTree.getByText("in/").nth(1)).toBeVisible();
+    await expect(sourceTree.getByText("git/")).toBeVisible();
+    await expect(sourceTree.getByText("repositories/")).toBeVisible();
+    await expect(sourceTree.getByText(".gitkeep")).toBeVisible();
+
+    await expect(
+      page.locator("text=0801ace Add a deeply nested directory tree"),
+    ).toBeVisible();
+  }
+});
+
+test("files with special characters in the filename", async ({ page }) => {
+  await page.goto(sourceBrowsingFixture);
+
+  const sourceTree = page.locator(".source-tree");
+  await sourceTree.getByText("special/").click();
+
+  await sourceTree.getByText("+plus+").click();
+  await expect(page.locator(".file-name")).toContainText("+plus");
+
+  await sourceTree.getByText("-dash-").click();
+  await expect(page.locator(".file-name")).toContainText("-dash-");
+
+  await sourceTree.getByText(":colon:").click();
+  await expect(page.locator(".file-name")).toContainText(":colon:");
+
+  await sourceTree.getByText(";semicolon;").click();
+  await expect(page.locator(".file-name")).toContainText(";semicolon;");
+
+  await sourceTree.getByText("@at@").click();
+  await expect(page.locator(".file-name")).toContainText("@at@");
+
+  await sourceTree.getByText("_underscore_").click();
+  await expect(page.locator(".file-name")).toContainText("_underscore_");
+
+  // TODO: fix the error in the app for these edge cases.
```

Comment:
I would probably change this to `fix the error in the http-api` since the app doesn't have any issues with these values

Reply:
695fcdeaae09d24622e138096ba2913bf48140f3